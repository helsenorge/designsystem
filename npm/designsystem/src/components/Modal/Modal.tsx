import React, { useEffect } from 'react';
import cn from 'classnames';

import { palette } from '../../theme/palette';
import Button from '../Button';
import Icon, { IconSize } from '../Icons';

import styles from './styles.module.scss';
import AlertSignStroke from '../Icons/AlertSignStroke';
import AlertSignFill from '../Icons/AlertSignFill';
import useFocusTrap from '../../hooks/useFocusTrap';
import { useIsVisible } from '../../hooks/useIsVisible';
import Title from '../Title/Title';
import { uuid } from '../../utils/uuid';
import Close from '../Close';
import CheckOutline from '../Icons/CheckOutline';
import Portal from '../Portal';
import { AnalyticsId, ZIndex } from '../../constants';

export enum ModalVariants {
  normal = 'normal',
  warning = 'warning',
  error = 'error',
  success = 'success',
  image = 'image',
}

export enum ModalSize {
  large = 'large',
  medium = 'medium',
}

export interface ModalProps {
  /** Title of the modal */
  title: string;
  /** id of the modal title */
  titleId?: string;
  /** Description of the modal. Will not render if the modal has children. */
  description?: string;
  /** Changes the visual representation of the modal */
  variant?: keyof typeof ModalVariants;
  /** Change width of the modal (default: large) */
  size?: keyof typeof ModalSize;
  /** Icon displayed in title */
  icon?: React.ReactElement;
  /** Hides the close button */
  noCloseButton?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Primary button text */
  primaryButtonText?: string;
  /** Secondary button text */
  secondaryButtonText?: string;
  /** Sets the aria-label of the modal */
  ariaLabel?: string;
  /** Sets the aria-labelledby of the modal */
  ariaLabelledBy?: string;
  /** Close button aria-label */
  ariaLabelCloseBtn?: string;
  /** Alternative component to modal */
  children?: React.ReactNode;
  /** Component shown after title */
  afterTitleChildren?: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Customize the z-index of the modal */
  zIndex?: number;
  /** Function is called when user clicks primary button */
  onSuccess?: () => void;
  /** Function is called when user clicks secondary button, clicks escape or outside the modal */
  onClose?: () => void;
  /** When enabled the component will be rendered in the bottom of document.body */
  printable?: boolean;
  /** If disabled, clicking escape or outside the modal will not close it */
  disableCloseEvents?: boolean;
}

const defaultProps = {
  variant: ModalVariants.normal,
  primaryButtonText: 'OK',
  titleId: uuid(),
  className: '',
  size: ModalSize.large,
  zIndex: ZIndex.Modal,
};

const getVariantIcon = (variant?: ModalProps['variant']): JSX.Element | null => {
  if (variant === ModalVariants.error) {
    return <Icon size={IconSize.Small} svgIcon={AlertSignFill} color={palette.cherry500} />;
  } else if (variant === ModalVariants.warning) {
    return <Icon size={IconSize.Small} svgIcon={AlertSignStroke} color={palette.black} />;
  } else if (variant === ModalVariants.success) {
    return <Icon size={IconSize.Small} svgIcon={CheckOutline} color={palette.kiwi900} />;
  }
  return null;
};

const getIcon = (variant?: ModalProps['variant'], icon?: ModalProps['icon']): JSX.Element | null => {
  const variantIcon = getVariantIcon(variant);
  if (variantIcon) {
    return <div className={styles.modal__iconWrapper}>{variantIcon}</div>;
  }
  if (icon) {
    return (
      <div className={styles.modal__iconWrapper}>
        {React.cloneElement(icon, {
          size: IconSize.Small,
        })}
      </div>
    );
  }
  return null;
};

const Modal = (props: ModalProps): JSX.Element => {
  const topContent = React.useRef<HTMLDivElement>(null);
  const modalContentRef = React.useRef<HTMLDivElement>(null);
  const dialogRef = React.useRef<HTMLDivElement>(null);
  useFocusTrap(dialogRef, true, true);
  const topContentVisible = useIsVisible(topContent);
  const bottomContent = React.useRef<HTMLDivElement>(null);
  const bottomContentVisible = useIsVisible(bottomContent);
  const contentIsScrollable = modalContentRef.current && modalContentRef.current.scrollHeight > modalContentRef.current.clientHeight;

  function handleKeyboardEvent(e: KeyboardEvent): void {
    if (e.key === 'Escape' && props.onClose) {
      e.stopPropagation();
      props.onClose();
    }
  }

  function handleClickEvent(event: MouseEvent): void {
    if (event.target && overlayRef.current === event.target && props.onClose) {
      event.stopPropagation();
      props.onClose();
    }
  }

  function disableBodyScroll(): void {
    document.body.style.overflow = 'hidden';
  }

  function enableBodyScroll(): void {
    document.body.style.removeProperty('overflow');
  }

  /* Displays a full window size modal with image */
  const imageView = props.variant === ModalVariants.image;

  const overlayRef = React.useRef<HTMLDivElement>(null);

  const showActions = (props.secondaryButtonText && props.secondaryButtonText?.length > 0) || props.onSuccess;

  // ariaLabelledBy prioriteres over ariaLabel, men dersom ariaLabel brukes trengs ikke ariaLabelledBy
  const ariaLabel = !props.ariaLabelledBy ? props.ariaLabel : undefined;
  const ariaLabelledBy = props.ariaLabelledBy ? props.ariaLabelledBy : !props.ariaLabel ? props.titleId : undefined;

  useEffect(() => {
    const overlayElement = overlayRef.current;
    disableBodyScroll();
    if (!props.disableCloseEvents && overlayElement) {
      overlayElement.addEventListener('keydown', handleKeyboardEvent);
      overlayElement.addEventListener('click', handleClickEvent);
    }
    return (): void => {
      enableBodyScroll();
      if (!props.disableCloseEvents && overlayElement) {
        overlayElement.removeEventListener('keydown', handleKeyboardEvent);
        overlayElement.removeEventListener('click', handleClickEvent);
      }
    };
  }, [props.disableCloseEvents]);

  const dialogClasses = cn(
    props.className,
    styles.modal,
    styles[`modal--${props.variant}`],
    styles[`modal--${props.size}`],
    contentIsScrollable && !showActions && styles['modal--no-actions']
  );

  const titleClasses = cn({
    [styles['modal__title--error']]: props.variant === ModalVariants.error,
    [styles['modal__title--success']]: props.variant === ModalVariants.success,
  });

  const Component = (
    <div data-testid="dialog-container">
      <div
        ref={overlayRef}
        className={styles['modal-overlay']}
        data-testid={props.testId}
        data-analyticsid={AnalyticsId.Modal}
        style={{ zIndex: props.zIndex }}
      >
        <div className={styles.align}>
          <div className={dialogClasses} role="dialog" aria-label={ariaLabel} aria-labelledby={ariaLabelledBy} ref={dialogRef}>
            <div
              className={cn(styles['modal__shadow'], styles['modal__shadow--top'], {
                [styles['modal__shadow--show']]: !topContentVisible && contentIsScrollable,
              })}
            />
            <div
              className={cn(styles.modal__contentWrapper, {
                [styles['modal__contentWrapper--image']]: imageView,
              })}
              ref={modalContentRef}
            >
              {!props.noCloseButton && (
                <div className={styles.modal__closeWrapper}>
                  <div className={cn(styles.modal__closeWrapper__close)}>
                    <Close onClick={props.onClose} ariaLabel={props.ariaLabelCloseBtn} />
                  </div>
                </div>
              )}
              <div
                className={cn(styles[`modal__contentWrapper__scroll--${props.size}`], {
                  [styles['modal__contentWrapper__scroll--image']]: imageView,
                })}
              >
                <div ref={topContent} />
                <div className={styles.modal__contentWrapper__title}>
                  {getIcon(props.variant, props.icon)}
                  <Title id={ariaLabelledBy} htmlMarkup="h3" appearance="title3" className={titleClasses}>
                    {props.title}
                  </Title>
                  {props.afterTitleChildren && <div className={styles['modal__afterTitleChildren']}>{props.afterTitleChildren}</div>}
                </div>
                {imageView && (
                  <div>
                    <div className={styles['modal__contentWrapper__imageWrapper']}>{props.children}</div>
                    <span className={styles['modal__contentWrapper__imageDescription']}>{props.description}</span>
                  </div>
                )}
                {!imageView && props.children && <div>{props.children}</div>}
                {!imageView && !props.children && <p className={styles.modal__description}>{props.description}</p>}
                <div ref={bottomContent} />
              </div>
            </div>
            <div
              className={cn(styles['modal__shadow'], styles['modal__shadow--bottom'], {
                [styles['modal__shadow--show']]: !bottomContentVisible && contentIsScrollable,
              })}
            />
            {showActions && (
              <div className={cn(styles['modal__call-to-action'], styles[`modal__call-to-action--${props.size}`])}>
                {props.onSuccess && <Button onClick={props.onSuccess}>{props.primaryButtonText}</Button>}
                {props.secondaryButtonText && props.secondaryButtonText?.length > 0 && (
                  <Button variant="borderless" onClick={props.onClose}>
                    {props.secondaryButtonText}
                  </Button>
                )}
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );

  if (props.printable) {
    const printModal = 'print-modal';
    return (
      <Portal className={printModal} testId="print-modal">
        <style media="print">{`body > *:not(.${printModal}) {display: none;}`}</style>
        {Component}
      </Portal>
    );
  }

  return Component;
};

Modal.defaultProps = defaultProps;

export default Modal;
