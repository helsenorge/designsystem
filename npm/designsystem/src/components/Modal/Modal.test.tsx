import '@testing-library/jest-dom';
import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import Modal, { ModalVariants } from './Modal';
import userEvent from '@testing-library/user-event';

describe('Gitt at en modal skal vises ', (): void => {
  describe('Når en modal skal vise kun tittel og lukkeknapp', (): void => {
    it('Så skal den kun vise sist nevnte', (): void => {
      const onClose = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} />);
      const title = screen.getByText('Hei der');
      expect(title).toBeVisible();

      const buttons = screen.getAllByRole('button');
      expect(buttons).toHaveLength(1);
    });
  });
  describe('Når en bruker trykker på lukkeknappen', (): void => {
    it('Så skal onClose kalles en gang', (): void => {
      const onClose = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} />);

      const lukk = screen.getByRole('button');
      fireEvent.click(lukk);

      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når modal ikke skal ha lukkeknapp', (): void => {
    it('Så vises ikke lukkeknappen', (): void => {
      render(<Modal title="Hei der" noCloseButton />);

      const lukk = screen.queryByRole('button');
      expect(lukk).not.toBeInTheDocument();
    });
  });

  describe('Når modal skal ha en komponent etter tittelen', (): void => {
    it('Så vises komponenten', (): void => {
      render(<Modal title="Hei der" afterTitleChildren={<h2>{'Hei til deg også'}</h2>} />);

      const title = screen.queryByRole('heading', { name: 'Hei til deg også' });
      expect(title).toBeVisible();
    });
  });

  describe('Når modal har custom z-index', (): void => {
    it('Så er z-index satt med inline styling', (): void => {
      render(<Modal title="Hei der" zIndex={123} testId="modal-med-custom-z-index" />);

      const modal = screen.getByTestId('modal-med-custom-z-index');
      expect(modal).toHaveAttribute('style', 'z-index: 123;');
    });
  });

  describe('Når en bruker trykker på OK knappen', (): void => {
    it('Så skal onSuccess kalles en gang', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} onSuccess={onSuccess} />);

      const buttons = screen.getAllByRole('button');
      expect(buttons).toHaveLength(2);

      const okKnapp = screen.getByText('OK');
      fireEvent.click(okKnapp);

      expect(onClose).toHaveBeenCalledTimes(0);
      expect(onSuccess).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når en bruker trykker på Avbryt knappen', (): void => {
    it('Så skal onSuccess kalles en gang', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} onSuccess={onSuccess} secondaryButtonText="Avbryt" />);

      const buttons = screen.getAllByRole('button');
      expect(buttons).toHaveLength(3);

      const cancel = screen.getByText('Avbryt');
      fireEvent.click(cancel);

      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onSuccess).toHaveBeenCalledTimes(0);
    });
  });

  describe(`Når en modal har variant ${ModalVariants.warning}`, (): void => {
    it('Så skal bakgrunnen endre seg til riktig farge og det vises et ikon', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      const { container } = render(
        <Modal
          title="Hei der"
          titleId={'titleid01'}
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.warning}
        />
      );

      const dialog = screen.getByRole('dialog');
      expect(dialog.className).toBe(`modal modal--${ModalVariants.warning} modal--large`);

      expect(container).toMatchSnapshot();
    });
  });

  describe(`Når en modal har variant ${ModalVariants.error}`, (): void => {
    it('Så skal bakgrunnen endre seg til riktig farge og det vises et ikon', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      const { container } = render(
        <Modal
          title="Hei der"
          titleId={'titleid02'}
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
        />
      );

      const dialog = screen.getByRole('dialog');
      expect(dialog.className).toBe(`modal modal--${ModalVariants.error} modal--large`);

      expect(container).toMatchSnapshot();
    });
  });

  describe(`Når en modal har satt attributt large`, (): void => {
    it('så skal dialogen ha klassen large', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} onSuccess={onSuccess} secondaryButtonText="Avbryt" large />);

      const dialog = screen.getByRole('dialog');
      expect(dialog.className).toBe('modal modal--normal modal--large');
    });
  });

  describe(`Når en modal har satt attributt imageView`, (): void => {
    it('så skal dialogen ha klassen imageView', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(<Modal variant={ModalVariants.image} title="Hei der" onClose={onClose} onSuccess={onSuccess} secondaryButtonText="Avbryt" />);

      const dialog = screen.getByRole('dialog');
      expect(dialog.className).toBe('modal modal--image modal--large');
    });
  });

  describe(`Når en modal har satt attributt imageView og large`, (): void => {
    it('så skal dialogen ha klassen imageView', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal variant={ModalVariants.image} title="Hei der" onClose={onClose} onSuccess={onSuccess} secondaryButtonText="Avbryt" large />
      );

      const dialog = screen.getByRole('dialog');
      expect(dialog.className).toBe('modal modal--image modal--large');
    });
  });

  describe(`Når en bruker tabber flere ganger inne i en Modal`, (): void => {
    it('Så skal man fanges inne i modalen når man tabber', async (): Promise<void> => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <>
          <Modal
            title="Hei der"
            onClose={onClose}
            onSuccess={onSuccess}
            secondaryButtonText="Avbryt"
            variant={ModalVariants.error}
            testId="testid"
          />
        </>
      );

      const buttons = screen.getAllByRole('button');
      expect(buttons[0]).toHaveFocus();
      await userEvent.tab();
      expect(buttons[1]).toHaveFocus();
      await userEvent.tab();
      expect(buttons[2]).toHaveFocus();
      await userEvent.tab();
      expect(buttons[0]).toHaveFocus();
    });
  });

  describe(`Når en modal har knapperad nederst (CTA), og en bruker trykker utenfor modalen`, (): void => {
    it('Så skal modalen ikke lukkes', async (): Promise<void> => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal
          title="Hei der"
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
          testId="testid"
        />
      );

      const dialog = screen.getByTestId('testid');

      await userEvent.click(dialog);

      expect(onClose).toHaveBeenCalled();
    });
  });

  describe(`Når en modal ikke har knapperad nederst (CTA), og en bruker trykker utenfor modalen`, (): void => {
    it('Så skal modalen lukkes', async (): Promise<void> => {
      const onClose = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} variant={ModalVariants.error} testId="testid" />);

      const dialog = screen.getByTestId('testid');

      await userEvent.click(dialog);

      expect(onClose).toHaveBeenCalled();
    });
  });

  describe(`Når en modal har knapperad nederst (CTA), og en bruker trykker Escape`, (): void => {
    it('Så skal modalen lukkes', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal
          title="Hei der"
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
          testId="testid"
        />
      );

      const dialog = screen.getByTestId('testid');

      fireEvent.keyDown(dialog, { key: 'Escape' });

      expect(onClose).toBeCalled();
    });
  });

  describe(`Når en modal ikke har en knapperad (CTA) nederst og en bruker trykker Escape`, (): void => {
    it('Så skal modalen lukkes', (): void => {
      const onClose = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} variant={ModalVariants.error} testId="testid" />);

      const dialog = screen.getByTestId('testid');

      fireEvent.keyDown(dialog, { key: 'Escape' });

      expect(onClose).toBeCalled();
    });
  });

  describe(`Når en modal ikke har en knapperad (CTA) nederst, og har disableCloseEvents til true`, (): void => {
    it('Så skal Escape ikke få modalen til å lukkes', (): void => {
      const onClose = jest.fn();

      render(<Modal title="Hei der" onClose={onClose} variant={ModalVariants.error} disableCloseEvents testId="testid" />);

      const dialog = screen.getByTestId('testid');

      fireEvent.keyDown(dialog, { key: 'Escape' });

      expect(onClose).not.toBeCalled();
    });
  });

  describe(`Når en modal rendres med titleId`, (): void => {
    it('Så skal titleId settes', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal
          title="Hei der"
          titleId="titleIdTest"
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
          testId="testid"
        />
      );

      const title = screen.getByText('Hei der');

      expect(title.id).toBe('titleIdTest');
    });
  });

  describe(`Når en modal rendres med ariaLabel satt`, (): void => {
    it('Så skal ariaLabel settes', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal
          title="Hei der"
          ariaLabel={'aria label test'}
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
          testId="testid"
        />
      );

      const dialog = screen.getByLabelText('aria label test');

      expect(dialog).toBeVisible();
    });
  });

  describe(`Når en modal rendres med ariaLabelledBy satt`, (): void => {
    it('Så skal ariaLabel settes etter ariaLabbeledBy og ikke ariaLabel', (): void => {
      const onClose = jest.fn();
      const onSuccess = jest.fn();

      render(
        <Modal
          title="Hei der"
          titleId="titleId"
          ariaLabel={'aria label test'}
          ariaLabelledBy={'titleId'}
          onClose={onClose}
          onSuccess={onSuccess}
          secondaryButtonText="Avbryt"
          variant={ModalVariants.error}
          testId="testid"
        />
      );

      const dialog = screen.getByLabelText('Hei der');

      expect(dialog).toBeVisible();
    });
  });

  describe('Når modal skal vises med custom titleId', (): void => {
    test('Så har tittelen id-en', (): void => {
      render(<Modal title="Dette er tittelen" titleId="testing" />);

      const label = screen.getByText('Dette er tittelen');
      expect(label).toHaveAttribute('id', 'testing');
    });
  });
});

describe('Gitt at en modal skal vises som printable ', (): void => {
  it('Så skal modalen wrappes med Portal', (): void => {
    const onClose = jest.fn();
    const onSuccess = jest.fn();

    render(
      <Modal
        title="Hei der"
        description="Dette er innhold"
        onClose={onClose}
        onSuccess={onSuccess}
        secondaryButtonText="Avbryt"
        variant={ModalVariants.error}
        testId="testid"
        titleId="test"
        printable
      />
    );
    const portal = screen.getByTestId('print-modal');
    expect(portal).toBeVisible();
    expect(portal.innerHTML).toMatchSnapshot();

    const dialog = screen.getByLabelText('Hei der');
    expect(dialog).toBeVisible();
  });
});
