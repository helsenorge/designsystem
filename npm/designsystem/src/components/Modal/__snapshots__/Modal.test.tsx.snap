// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at en modal skal vises  Når en modal har variant error Så skal bakgrunnen endre seg til riktig farge og det vises et ikon 1`] = `
<div>
  <div
    data-testid="dialog-container"
  >
    <div
      class="modal-overlay"
      data-analyticsid="modal"
      style="z-index: 1300000;"
    >
      <div
        class="align"
      >
        <div
          aria-labelledby="titleid02"
          class="modal modal--error modal--large"
          role="dialog"
        >
          <div
            class="modal__shadow modal__shadow--top"
          />
          <div
            class="modal__contentWrapper"
          >
            <div
              class="modal__closeWrapper"
            >
              <div
                class="modal__closeWrapper__close"
              >
                <button
                  aria-label="Lukk"
                  class="close"
                  data-analyticsid="close"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="hnds-style-icon"
                    data-analyticsid="icon"
                    fill="#126F87"
                    focusable="false"
                    height="48"
                    role="presentation"
                    style="min-width: 48px; min-height: 48px;"
                    viewBox="0 0 48 48"
                    width="48"
                  >
                    <path
                      d="M25.403 24l10.259-10.259-1.403-1.403L24 22.597l-10.259-10.26-1.403 1.403L22.597 24 12.338 34.26l1.403 1.403L24 25.403l10.259 10.259 1.403-1.403z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="modal__contentWrapper__scroll--large"
            >
              <div />
              <div
                class="modal__contentWrapper__title"
              >
                <div
                  class="modal__iconWrapper"
                >
                  <svg
                    aria-hidden="true"
                    class="hnds-style-icon"
                    data-analyticsid="icon"
                    fill="#C83521"
                    focusable="false"
                    height="48"
                    role="presentation"
                    style="min-width: 48px; min-height: 48px;"
                    viewBox="0 0 48 48"
                    width="48"
                  >
                    <path
                      d="M24.898 21.447v7.228a.893.893 0 01-1.785 0v-7.228a.892.892 0 111.785 0zm.5 11.601a1.394 1.394 0 11-2.787-.001 1.394 1.394 0 012.787.001zm-18.822 6.71h34.847L24.111 8.242 6.576 39.758z"
                    />
                  </svg>
                </div>
                <h3
                  class="title title--title3 modal__title--error"
                  data-analyticsid="title"
                  id="titleid02"
                  style="margin-top: 0rem; margin-bottom: 0rem;"
                >
                  Hei der
                </h3>
              </div>
              <p
                class="modal__description"
              />
              <div />
            </div>
          </div>
          <div
            class="modal__shadow modal__shadow--bottom"
          />
          <div
            class="modal__call-to-action modal__call-to-action--large"
          >
            <button
              class="button-wrapper"
              data-analyticsid="button"
              type="button"
            >
              <div
                class="button button--normal"
              >
                <div
                  class="button__text"
                >
                  <span>
                    OK
                  </span>
                </div>
              </div>
            </button>
            <button
              class="button-wrapper"
              data-analyticsid="button"
              type="button"
            >
              <div
                class="button button--normal button--borderless"
              >
                <div
                  class="button__text"
                >
                  <span>
                    Avbryt
                  </span>
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Gitt at en modal skal vises  Når en modal har variant warning Så skal bakgrunnen endre seg til riktig farge og det vises et ikon 1`] = `
<div>
  <div
    data-testid="dialog-container"
  >
    <div
      class="modal-overlay"
      data-analyticsid="modal"
      style="z-index: 1300000;"
    >
      <div
        class="align"
      >
        <div
          aria-labelledby="titleid01"
          class="modal modal--warning modal--large"
          role="dialog"
        >
          <div
            class="modal__shadow modal__shadow--top"
          />
          <div
            class="modal__contentWrapper"
          >
            <div
              class="modal__closeWrapper"
            >
              <div
                class="modal__closeWrapper__close"
              >
                <button
                  aria-label="Lukk"
                  class="close"
                  data-analyticsid="close"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="hnds-style-icon"
                    data-analyticsid="icon"
                    fill="#126F87"
                    focusable="false"
                    height="48"
                    role="presentation"
                    style="min-width: 48px; min-height: 48px;"
                    viewBox="0 0 48 48"
                    width="48"
                  >
                    <path
                      d="M25.403 24l10.259-10.259-1.403-1.403L24 22.597l-10.259-10.26-1.403 1.403L22.597 24 12.338 34.26l1.403 1.403L24 25.403l10.259 10.259 1.403-1.403z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="modal__contentWrapper__scroll--large"
            >
              <div />
              <div
                class="modal__contentWrapper__title"
              >
                <div
                  class="modal__iconWrapper"
                >
                  <svg
                    aria-hidden="true"
                    class="hnds-style-icon"
                    data-analyticsid="icon"
                    fill="#000000"
                    focusable="false"
                    height="48"
                    role="presentation"
                    style="min-width: 48px; min-height: 48px;"
                    viewBox="0 0 48 48"
                    width="48"
                  >
                    <path
                      d="M24.003 31.578a1.394 1.394 0 100 2.788 1.394 1.394 0 000-2.788zm.098-20.19l14.687 26.735H9.227L24.1 11.388zm.008-2.929L6.816 39.541h34.367L24.109 8.459zm-.106 21.033c.494 0 .893-.4.893-.893v-7.228a.893.893 0 00-1.786 0v7.228c0 .493.4.893.893.893z"
                    />
                  </svg>
                </div>
                <h3
                  class="title title--title3"
                  data-analyticsid="title"
                  id="titleid01"
                  style="margin-top: 0rem; margin-bottom: 0rem;"
                >
                  Hei der
                </h3>
              </div>
              <p
                class="modal__description"
              />
              <div />
            </div>
          </div>
          <div
            class="modal__shadow modal__shadow--bottom"
          />
          <div
            class="modal__call-to-action modal__call-to-action--large"
          >
            <button
              class="button-wrapper"
              data-analyticsid="button"
              type="button"
            >
              <div
                class="button button--normal"
              >
                <div
                  class="button__text"
                >
                  <span>
                    OK
                  </span>
                </div>
              </div>
            </button>
            <button
              class="button-wrapper"
              data-analyticsid="button"
              type="button"
            >
              <div
                class="button button--normal button--borderless"
              >
                <div
                  class="button__text"
                >
                  <span>
                    Avbryt
                  </span>
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Gitt at en modal skal vises som printable  Så skal modalen wrappes med Portal 1`] = `"<style media="print">body > *:not(.print-modal) {display: none;}</style><div data-testid="dialog-container"><div class="modal-overlay" data-testid="testid" data-analyticsid="modal" style="z-index: 1300000;"><div class="align"><div class="modal modal--error modal--large" role="dialog" aria-labelledby="test"><div class="modal__shadow modal__shadow--top"></div><div class="modal__contentWrapper"><div class="modal__closeWrapper"><div class="modal__closeWrapper__close"><button data-analyticsid="close" class="close" aria-label="Lukk" type="button"><svg data-analyticsid="icon" class="hnds-style-icon" role="presentation" focusable="false" aria-hidden="true" viewBox="0 0 48 48" style="min-width: 48px; min-height: 48px;" width="48" height="48" fill="#126F87"><path d="M25.403 24l10.259-10.259-1.403-1.403L24 22.597l-10.259-10.26-1.403 1.403L22.597 24 12.338 34.26l1.403 1.403L24 25.403l10.259 10.259 1.403-1.403z"></path></svg></button></div></div><div class="modal__contentWrapper__scroll--large"><div></div><div class="modal__contentWrapper__title"><div class="modal__iconWrapper"><svg data-analyticsid="icon" class="hnds-style-icon" role="presentation" focusable="false" aria-hidden="true" viewBox="0 0 48 48" style="min-width: 48px; min-height: 48px;" width="48" height="48" fill="#C83521"><path d="M24.898 21.447v7.228a.893.893 0 01-1.785 0v-7.228a.892.892 0 111.785 0zm.5 11.601a1.394 1.394 0 11-2.787-.001 1.394 1.394 0 012.787.001zm-18.822 6.71h34.847L24.111 8.242 6.576 39.758z"></path></svg></div><h3 id="test" class="title title--title3 modal__title--error" style="margin-top: 0rem; margin-bottom: 0rem;" data-analyticsid="title">Hei der</h3></div><p class="modal__description">Dette er innhold</p><div></div></div></div><div class="modal__shadow modal__shadow--bottom"></div><div class="modal__call-to-action modal__call-to-action--large"><button data-analyticsid="button" class="button-wrapper" type="button"><div class="button button--normal"><div class="button__text"><span>OK</span></div></div></button><button data-analyticsid="button" class="button-wrapper" type="button"><div class="button button--normal button--borderless"><div class="button__text"><span>Avbryt</span></div></div></button></div></div></div></div></div>"`;
